### PR TITLE
Improve shutdown of non-persistent HTTP/1 connections #12212

### LINKED ISSUE #12212: ShutdownOutput for non-persistent HTTP/1 connections
**Jetty version(s)**
12

**Enhancement Description**

Currently a needless exception is thrown by the HttpConnection class after every connection is closed:
```
2024-08-30 12:00:37.908:DEBUG:oejsi.HttpConnection:qtp1212899836-30: caught exception HttpConnection@28edcae7::SocketChannelEndPoint@35c31367[{l=/127.0.0.1:8080,r=/127.0.0.1:33516,OPEN,fill=-,flush=-,to=3/30000}{io=0/0,kio=0,kro=1}]->[HttpConnection@28edcae7[p=HttpParser{s=CLOSE,0 of -1},g=HttpGenerator@7c03111a{s=START}]=>HttpChannelState@72c4f0ea{handling=null, handled=false, send=SENDING, completed=false, request=null}] HttpChannelState@72c4f0ea{handling=null, handled=false, send=SENDING, completed=false, request=null}
org.eclipse.jetty.io.RuntimeIOException: Parser is terminated
	at org.eclipse.jetty.server.internal.HttpConnection.parseRequestBuffer(HttpConnection.java:559)
	at org.eclipse.jetty.server.internal.HttpConnection.onFillable(HttpConnection.java:384)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:322)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:99)
	at org.eclipse.jetty.io.SelectableChannelEndPoint$1.run(SelectableChannelEndPoint.java:53)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.runTask(AdaptiveExecutionStrategy.java:478)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.consumeTask(AdaptiveExecutionStrategy.java:441)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.tryProduce(AdaptiveExecutionStrategy.java:293)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.produce(AdaptiveExecutionStrategy.java:195)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:979)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.doRunJob(QueuedThreadPool.java:1209)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1164)
	at java.base/java.lang.Thread.run(Thread.java:1570)
```

This can be seen if debug is turned on for the following test server 
```java

    public static void main(String... args) throws Exception
    {
        Server server = new Server();

        ServerConnector connector = new ServerConnector(server);
        server.addConnector(connector);
        connector.setPort(8080);
        connector.setIdleTimeout(30000);
        HttpConfiguration httpConfiguration = connector.getConnectionFactory(HttpConnectionFactory.class).getHttpConfiguration();
        httpConfiguration.setHttpCompliance(HttpCompliance.RFC2616);
        httpConfiguration.setIdleTimeout(30000);

        ContextHandler context = new ContextHandler();
        server.setHandler(context);

        context.setHandler(new Handler.Abstract()
        {
            @Override
            public boolean handle(Request request, Response response, Callback callback) throws Exception
            {
                byte[] bytes = "Now is time for all good men to come to the aid of the party!\n".getBytes(StandardCharsets.UTF_8);
                response.write(true, BufferUtil.toBuffer(bytes), callback);
                return true;
            }
        });

        server.start();
        server.join();
    }
```

and then requests like the following are sent:

```
ardesia: ~/src/jetty-12.0.x/jetty-home/target/jetty-home/base (jetty-12.0.x)
[1991] telnet localhost 8080
Trying 127.0.0.1...
Connected to localhost.
Escape character is '^]'.
GET /
Now is time for all good men to come to the aid of the party!
Connection closed by foreign host.

ardesia: ~/src/jetty-12.0.x/jetty-home/target/jetty-home/base (jetty-12.0.x)
[1991] telnet localhost 8080
Trying 127.0.0.1...
Connected to localhost.
Escape character is '^]'.
GET / HTTP/1.0

HTTP/1.1 200 OK
Server: Jetty(12.0.13-SNAPSHOT)
Date: Fri, 30 Aug 2024 01:53:41 GMT
Content-Length: 62

Now is time for all good men to come to the aid of the party!
Connection closed by foreign host.

ardesia: ~/src/jetty-12.0.x/jetty-home/target/jetty-home/base (jetty-12.0.x)
[1991] telnet localhost 8080
Trying 127.0.0.1...
Connected to localhost.
Escape character is '^]'.
GET / HTTP/1.1
Host: localhost:8080
Connection: close

HTTP/1.1 200 OK
Server: Jetty(12.0.13-SNAPSHOT)
Date: Fri, 30 Aug 2024 01:54:18 GMT
Content-Length: 62
Connection: close

Now is time for all good men to come to the aid of the party!
Connection closed by foreign host.
```

### PR DESCRIPTION
**Pull request by gregw**
Fix #12212 by improving shutdown of non-persistent HTTP/1 connections 
 + shutdown in SendCallback
 + keep failsafe in Stream callback
 + improve chunk toString clarity
 + added failsafe release in onFillable, to handle case when endpoint is closed.

### CODE REVIEW

**Comment by gregw** [comment]
@lachlan-roberts @sbordet @lorban This work builds on #12213 by moving the shutdownOutput to the SendCallback, rather than wait for the Stream callback completion (which may be long after the response is completed).

Whilst working on this, I noticed that onFillable could exit without releasing the request buffer. So I've added a failsafe release.    However, it would be good to review the release logic of the class as part of the review of this PR.
I also improved the chunk toString output, as debugging was difficult with the previous verbose toString.

I believe this change may help with the HTTP/0.9 issues we have seen

File: jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/internal/ByteBufferChunk.java [hunk 1]
```diff
@@ -169,7 +170,11 @@ public boolean release()
         @Override
         public String toString()
         {
-            return "%s[%s]".formatted(super.toString(), retainable);
+            Retainable r = retainable;
+            while (r instanceof Retainable.Wrapper wrapper)
+                r = wrapper.getWrapped();
+
+            return "Chunk.%s[%s]".formatted(super.toString(), r instanceof AbstractRetainableByteBuffer abstractBuffer ? abstractBuffer.toReferenceString() : r.toString());
```
Comment (sbordet) [inline jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/internal/ByteBufferChunk.java]: This should not be necessary, as the base class should print its own fields.
This should just be: `return "Chunk.%s[%s]".formatted(super.toString(), r);`

File: jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/AbstractRetainableByteBuffer.java [hunk 2]
```diff
@@ -75,4 +75,10 @@ public String toString()
     {
         return "%s@%x[rc=%d,%s]".formatted(getClass().getSimpleName(), hashCode(), refCount.get(), BufferUtil.toDetailString(byteBuffer));
     }
+
+    public String toReferenceString()
```
Comment (sbordet) [inline jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/AbstractRetainableByteBuffer.java]: Is this really necessary to be exposed publicly?
It is already printed in the `toString()` above.
Actually, I don't think it is necessary.

File: jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpConnection.java [hunk 3]
```diff
@@ -341,9 +342,9 @@ void releaseRequestBuffer()
         {
             if (LOG.isDebugEnabled())
                 LOG.debug("releaseRequestBuffer {}", this);
-            if (_retainableByteBuffer.release())
-                _retainableByteBuffer = null;
-            else
+            RetainableByteBuffer buffer = _retainableByteBuffer;
+            _retainableByteBuffer = null;
+            if (!buffer.release())
                 throw new IllegalStateException("unreleased buffer " + _retainableByteBuffer);
```
Comment (sbordet) [inline jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpConnection.java]: This will now print `unreleased buffer null`, should use the `buffer` local variable, not the field.

File: jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpConnection.java [hunk 4]
```diff
@@ -457,6 +458,13 @@ else if (filled == 0)
         finally
         {
             setCurrentConnection(last);
+
+            // If handling has completed
+            if (_stream.get() == null && _retainableByteBuffer != null)
+            {
+                _retainableByteBuffer.release();
+                _retainableByteBuffer = null;
+            }
```
Comment (sbordet) [inline jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpConnection.java]: This seems fishy.

The usual pattern is to release the buffer when a `fill()` returned `0` or `-1`, where we know we are not going to use the buffer.

Doing it in the finally is in a race condition, as when `filled==0` we set fill interest, and another thread could re-enter, use the current buffer that is actually being released, so not good.

The buffer should be properly handled in `fillRequestBuffer()`. What prompted you to add this?

File: jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpConnection.java [hunk 5]
```diff
@@ -906,12 +914,16 @@ private void releaseChunk()
         @Override
         protected void onCompleteSuccess()
         {
+            if (_shutdownOut && !(_httpChannel.getRequest().getAttribute(HttpStream.UPGRADE_CONNECTION_ATTRIBUTE) instanceof Connection))
+                getEndPoint().shutdownOutput();
```
Comment (sbordet) [inline jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpConnection.java]: Why is this needed here?
It is already taken care of in `HttpStreamOverHTTP1.succeeded()`.

File: jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpConnection.java [hunk 6]
```diff
@@ -906,12 +914,16 @@ private void releaseChunk()
...
             release().succeeded();
         }
 
         @Override
         public void onCompleteFailure(final Throwable x)
         {
+            if (_shutdownOut)
+                getEndPoint().shutdownOutput();
```
Comment (sbordet) [inline jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpConnection.java]: I think this should be moved to `HttpStreamOverHTTP1.failed()`.

Here it seems way too early, as a failure here may still allow to write an error response.

File: jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpConnection.java [hunk 7]
```diff
@@ -1523,13 +1534,12 @@ public void succeeded()
                 return;
             }
 
-            // As this is not an upgrade, we can shutdown the output if we know we are not persistent
-            if (_sendCallback._shutdownOut)
-                getEndPoint().shutdownOutput();
+            // Failsafe shutdown (should have been done by SendCallback)
+            if (_sendCallback._shutdownOut && !getEndPoint().isOutputShutdown())
+                IO.close(getEndPoint());
```
Comment (sbordet) [inline jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpConnection.java]: I don't think this is right. 

Below we even set fill interest to seek EOF, so we cannot close the endpoint here.

**Review by sbordet — CHANGES_REQUESTED** [review]
@gregw I fail to exactly understand what this PR is fixing.

Is there a new test case to be added, or an existing one that fails?

**Comment by gregw** [comment]
> @gregw I fail to exactly understand what this PR is fixing.

Did you read #12212 where it says:

```
In previous versions of jetty, we did the shutdown as the last write completed. We are now doing it as the handling callback is completed, which may be very much later than when the response is sent.
I think we should do the shutdown as soon as possible.
@sbordet thoughts?
```
 
So this is really just reverting to the behaviour that we've had for decades.  

> Is there a new test case to be added, or an existing one that fails?

I've written a test to check that the shutdown by a write(last) is not delayed until after callback.succeeded on the handle callback.

File: jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/internal/ByteBufferChunk.java (hunk 1 above)
Comment (gregw) [reply jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/internal/ByteBufferChunk.java]: The problem is that often the Retainable can be just the buffer itself, so you get the whole long buffer details twice.

File: jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/AbstractRetainableByteBuffer.java (hunk 2 above)
Comment (gregw) [reply jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/AbstractRetainableByteBuffer.java]: It is not strictly necessary, but if the RBB is being used just as a retainable, then I'd like a way to avoid printing the detail of the buffer twice in the toString above.

File: jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpConnection.java (hunk 5 above)
Comment (gregw) [reply jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpConnection.java]: Please read issue and comments!

File: jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpConnection.java (hunk 4 above)
Comment (gregw) [reply jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpConnection.java]: Yes it is fishy, hence I wrote:
> Whilst working on this, I noticed that onFillable could exit without releasing the request buffer. So I've added a failsafe release. However, it would be good to review the release logic of the class as part of the review of this PR

Releasing in the finally here is a fail safe I put in that was necessary since the buffer is not released if the endpoint is closed whilst there is a thread in handling.    This works and we could go forward with just that fix, but I am asking should we review/cleanup the whole releasing behaviour of this class in this PR (or is that for another PR?)

File: jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpConnection.java (hunk 6 above)
Comment (gregw) [reply jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpConnection.java]: Please read the issue and comments.   
This is just reverting to how we have shutdown for decades.  I see no reason why we started deferring the shutdown in 12.0 so that it could be delayed by application code long after the response is committed.

File: jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpConnection.java (hunk 7 above)
Comment (gregw) [reply jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpConnection.java]: I'm not sure we need the failsafe at all.  I think the sendCallback is always called.

But I just wanted to be cautious and make sure it is closed.  Maybe a shutdown here would be enough or maybe no failsafe at all.

File: jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpConnection.java (hunk 4 above)
Comment (sbordet) [reply jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpConnection.java]: Releasing in the finally is definitely racy.

Perhaps the best would be using `while (true)`, with a check inside for `endPoint.isOpen()`, and if not, release the buffer and break the loop.

File: jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpConnection.java (hunk 6 above)
Comment (sbordet) [reply jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpConnection.java]: Yet, in case of failures, it may be possible to fail the write and still wanting to send an error response instead, for example for headers too large where the response is not committed as we fail before, and failing the channel callback would allow to send a 431, but not if we shutdown the endpoint here.

File: jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpConnection.java (hunk 7 above)
Comment (sbordet) [reply jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpConnection.java]: No fail safe.

However, we need to make sure the failure path is correct.

File: jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpConnection.java (hunk 4 above)
Comment (gregw) [reply jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpConnection.java]: Also note that once _stream.get() == null, there will never be another thread entering here.  As I said, it is a fail-safe to protect against the `while(getEndPoint().isOpen()` above becoming an unreleasing noop if the endpoint is closed.

File: jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/internal/ByteBufferChunk.java (hunk 1 above)
Comment (gregw) [reply jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/internal/ByteBufferChunk.java]: I solved it a different way.

File: jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/AbstractRetainableByteBuffer.java (hunk 2 above)
Comment (gregw) [reply jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/AbstractRetainableByteBuffer.java]: I solved it a different way.

File: jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpConnection.java [hunk 8]
```diff
@@ -369,7 +369,7 @@ public void onFillable()
         HttpConnection last = setCurrentConnection(this);
         try
         {
-            while (getEndPoint().isOpen())
+            while (true)
```
Comment (gregw) [inline jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpConnection.java:375]: This was wrong, because by looking past the request (as a Content.Source) to the endpoint, this loop might become a noop, even when the endpoint has been consumed and buffered.    The end of reading should only be indicated by reading -1

File: jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpConnection.java (hunk 6 above)
Comment (gregw) [reply jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpConnection.java]: I'm a bit torn by this one.   Prior to jetty-12, we always shutdown here.    However, the failure indicates that we failed to write last, so I agree a shutdown here is wrong.     I'm almost inclined to do a close/abort here instead... but will try doing nothing so that the app can try to catch and write something different (feels dodgy to try... but maybe).

File: jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpConnection.java (hunk 8 above)
Comment (lorban) [reply jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpConnection.java:375]: That should be stated in a comment.

File: jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpConnection.java [hunk 9]
```diff
@@ -457,6 +458,13 @@ else if (filled == 0)
         finally
         {
             setCurrentConnection(last);
+
+            // If handling has completed
+            if (_stream.get() == null && _retainableByteBuffer != null)
```
Comment (lorban) [inline jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpConnection.java]: This condition is tricky to grasp and gives a sense of fishiness at first sight, as it happened for @sbordet and myself too.

After carefully reading the explanations below, I believe this is correct although quite delicate so IMHO the following should be carefully stated in a comment:

 - when `_stream.get() == null`, there will never be another thread entering here
 - `while(getEndPoint().isOpen())` becoming an unreleasing noop if the endpoint is closed

File: jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpConnection.java (hunk 5 above)
Comment (lorban) [reply jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpConnection.java]: That reason should be stated in a comment.

File: jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/HttpServerTestBase.java [hunk 10]
```diff
@@ -1085,6 +1088,59 @@ public void testCloseWhileWriteBlocked() throws Exception
         }
     }
 
+    @Test
+    public void testCloseWhileCompletePending() throws Exception
```
Comment (lorban) [inline jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/HttpServerTestBase.java:1094]: This test clearly shows the major difference between `Response.write(last = true)` and `Callback.succeeded()`.

In my mind, it's always been "the last write sends everything to the client to make it figure out the exchange over" and "succeeding the callback tells the server the exchange is over" but apparently that got broken by accident.

Do we have some documentation speaking about this?

**Review by lorban — CHANGES_REQUESTED** [review]
Code wise, I believe this change is good.

But the biggest issue here is IMHO that the difference between the last write and succeeding the callback is a bit tricky to implement and may need to be better javadoc'ed.

File: jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpConnection.java (hunk 9 above)
Comment (gregw) [reply jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpConnection.java]: I'm going to defer the release fixes to another PR .

File: jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/HttpServerTestBase.java (hunk 10 above)
Comment (gregw) [reply jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/HttpServerTestBase.java:1094]: I think it is obvious in the contract.  I'm not sure why somebody (probably me) moved the shutdown to the wrong place.... must have been brain fade.

**Comment by gregw** [comment]
@sbordet @lorban fixing the buffer release in HttpConnection proved to be a game of whack-a-bug!  
So I have removed the majority of the release "fixes" from this PR and they need to be addressed in different one.   This PR now just moves the shutdown back to the sendCallback, but leaves a failsafe in the stream (apparently needed for some SSL tests).
I'd like to get this merged for this release and then look at the release issues next cycle.

The problems with release include:
 + the `releaseRequestBuffer` only does a conditional release of an empty buffer, when is some cases (e.g. failure, upgrade) we want to always release.
 + the `releaseRequestBuffer` method throws if the release is not the last one, which is counter to the concept of allowing chunks to be held with a retain.
 + when we release because the buffer has been retained, we do not release if it is not empty (e.g. if another request has been pipelined).  

I think it is all mostly working, but it is fragile to changes.  I think it needs a bit more time to review and perhaps a big catchall release called when the stream is completed (i.e. after both succeed/fail AND the last write has completed AND the handler thread has returned).

**Comment by sbordet** [comment]
@gregw I reduced to the minimum the changes necessary, and filed #12227 for further work.

Also, @lorban fixed a number of instabilities in #12226.

**Review by sbordet — APPROVED** [review]

**Review by lorban — APPROVED** [review]
